src/credentials/providers/credentialsProviderManager.ts

```diff
@@ -0,0 +1,60 @@
+/*!
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { CredentialsProvider } from './credentialsProvider'
+import { CredentialsProviderFactory } from './credentialsProviderFactory'
+import { makeCredentialsProviderIdComponents } from './credentialsProviderId'
+
+/**
+ * Responsible for providing the Toolkit with all available CredentialsProviders.
+ */
+export class CredentialsProviderManager {
+    private static INSTANCE: CredentialsProviderManager | undefined
+    private readonly providerFactories: CredentialsProviderFactory[] = []
+
+    public async getAllCredentialsProviders(): Promise<CredentialsProvider[]> {
+        const providers: CredentialsProvider[] = []
+
+        for (const factory of this.providerFactories) {
+            await factory.refresh()
+
+            providers.push(...factory.listProviders())
```

Comment:
See comment about list providers already making a copy

Reply:
nit: can we just use `.concat` instead then, I think that makes it more clear what it is doing